"rate limit exceeded" for datalad-installer --sudo ok git-annex -m datalad/git-annex:release

see e.g. https://github.com/DanielDent/git-annex-remote-rclone/actions/runs/3750292044/jobs/6370226146 for https://github.com/DanielDent/git-annex-remote-rclone/pull/59

```
File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/datalad_installer.py", line 2386, in get_latest_release_asset
    for release in self.paginate(f"[https://api.github.com/repos/{repo}/releases](https://api.github.com/repos/%7Brepo%7D/releases)"):
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/datalad_installer.py", line 2301, in paginate
    with self.get(url) as r:
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/datalad_installer.py", line 2292, in get
    with urlopen(req) as r:
         ^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/urllib/request.py", line 525, in open
    response = meth(req, response)
               ^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/urllib/request.py", line 634, in http_response
    response = self.parent.error(
               ^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/urllib/request.py", line 563, in error
    return self._call_chain(*args)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/urllib/request.py", line 496, in _call_chain
    result = func(*args)
             ^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error [40](https://github.com/DanielDent/git-annex-remote-rclone/actions/runs/3750292044/jobs/6370226146#step:5:41)3: rate limit exceeded
Error: Process completed with exit code 1.
```

apparently it is a "new" issue in this repo, and which is odd !  could we then accompany that crash with status about rate limits we have at that point like we do in cron job for tinuous on smaug 

```
curl -fsSL -H "Authorization: token $GITHUB_TOKEN" https://api.github.com/rate_limit 
```

to see if we do really?  or should we just retry few times?

--- a/src/datalad_installer.py
+++ b/src/datalad_installer.py
@@ -2304,8 +2304,12 @@ class GitHubClient:
     def get(self, url: str) -> Iterator[Any]:
         log.debug("HTTP request: GET %s", url)
         req = Request(url, headers=self.headers)
-        with urlopen(req) as r:
-            yield r
+        try:
+            with urlopen(req) as r:
+                yield r
+        except URLError as e:
+            raise_for_ratelimit(e, self.headers.get("Authorization"))
+            raise
 
     def getjson(self, url: str) -> Any:
         with self.get(url) as r:
@@ -2462,6 +2466,31 @@ class MethodNotSupportedError(Exception):
     pass
 
 
+def raise_for_ratelimit(e: URLError, auth: Optional[str]) -> None:
+    if isinstance(e, HTTPError) and e.code == 403:
+        try:
+            resp = json.load(e)
+        except Exception:
+            return
+        if "API rate limit exceeded" in resp.get("message", ""):
+            if auth is not None:
+                url = "https://api.github.com/rate_limit"
+                log.debug("HTTP request: GET %s", url)
+                req = Request(url, headers={"Authorization": auth})
+                with urlopen(req) as r:
+                    resp = json.load(r)
+                log.info(
+                    "GitHub rate limit exceeded; details:\n\n%s\n",
+                    textwrap.indent(json.dumps(resp, indent=4), " " * 4),
+                )
+            else:
+                raise RuntimeError(
+                    "GitHub rate limit exceeded and GITHUB_TOKEN not set;"
+                    " suggest setting GITHUB_TOKEN in order to get increased"
+                    " rate limit"
+                )
+
+
 def download_file(
     url: str, path: Union[str, os.PathLike], headers: Optional[Dict[str, str]] = None
 ) -> None:
@@ -2482,6 +2511,7 @@ def download_file(
             return
         except URLError as e:
             if isinstance(e, HTTPError) and e.code not in (500, 502, 503, 504):
+                raise_for_ratelimit(e, headers.get("Authorization"))
                 raise
             try:
                 delay = next(delays)

